Return a unique_ptr to message instead of void

satgera opened the pull request:
Summary:
Return a unique_ptr to message instead of void
This is to help thrift style rpc where there is no need for explicit send for a response.
If unique_ptr isn't intuitive fine with returning message itself. We need to figure out how to solve the non-blocking callback case but don't want to block the thrift based rpc agent implementation till then.

Differential Revision: D16825072

pietern commented:
Is the goal of the unique_ptr here to indicate if it's set or not? If so, I think you can use `c10::optional` as well and avoid a heap allocation. Otherwise, make sure to `std::move` the message into `make_unique` to avoid copying the whole object.

satgera committed:
Return a unique_ptr to message instead of void (#25253)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/25253

Return a unique_ptr to message instead of void
This is to help thrift style rpc where there is no need for explicit send for a response.
If unique_ptr isn't intuitive fine with returning message itself. We need to figure out how to solve the non-blocking callback case but don't want to block the thrift based rpc agent implementation till then.

Differential Revision: D16825072

fbshipit-source-id: 0034a516b0c90009e23936b9ca9bfb6d38dbdce3

mrshenli reviewed (commented):
Other than the @pietern's comments, this LGTM. Are you using `std::unique_ptr` instead of `c10::optional` to avoid forcing every `RpcAgent` implementation to depend on the entire libtorch? IIRC, you previous mentioned that it might be helpful to expose distributed or RPC as a separate lib to reduce the dependency size.
